lab-bundle-spec.md

```diff
@@ -308,20 +308,140 @@ environment:
+student_messages | ✓        | array of objects           | The keys are how the messages will be referenced in the code, and the values are locale dictionaries.
+services         | ✓        | array of resource services | An array of services that will be used in the code block. Each resource type specifies a set of allowed services.
+code             | ✓        | string                     | Code to be executed. See [below](#code) for more information.
+
+```yml
+assessment:
+  passing_percentage: 75
+  steps:
+    - title:
+        locales:
+          en: Create a Cloud Storage bucket
+          es: Crear un depósito de almacenamiento en la nube
+      maximum_score: 5
+      student_messages:
+        - success:
+            locales:
+              en:
+                Great job! You created the bucket!
+              es:
+                ¡Gran trabajo! ¡Creaste el cubo!
+        - bucket_missing:
+            locales:
+              en:
+                Oops! No bucket found.
+              es:
+                ¡Uy! No se ha encontrado el cubo.
+        - bucket_misconfigured:
+            locales:
+              en:
+                Hmm. The bucket is there, but it is misconfigured.
+              es:
+                Hmm. El cubo está allí, pero está mal configurado.
+      services:
+        - target_project.StorageV1
+      code: |-
+        def check(handles, max_score)
+          storage_handle = handles['target_project.StorageV1']
+
+          # Check for bucket
+          found_bucket = ...
+          unless found_bucket
+            return { score: 0, student_message: 'bucket_missing' }
+          end
+
+          # Check bucket configuration
+          bucket_configured_correctly = ...
+          unless bucket_configured_correctly
+            return { score: 2, student_message: 'bucket_misconfigured' }
+          end
+
+          { score: max_score, student_message: 'success' }
+        end
+    - title:
+        locales:
+          en: Copy a file to the bucket
+          es: Copiar un archivo al cubo
+      maximum_score: 5
+      student_messages:
+        - success:
+            locales:
+              en:
+                Great job! You copied the file!
+              es:
+                ¡Gran trabajo! ¡Copiaste el archivo!
+        - file_missing:
+            locales:
+              en:
+                Oops! No file found.
+              es:
+                ¡Uy! No se ha encontrado el archivo.
+        - file_mismatch:
+            locales:
+              en:
+                Hmm. There's a file here, but it doesn't match the source contents.
+              es:
+                Hmm. Hay un archivo aquí, pero no coincide con el contenido de origen.
+      services:
+        - source_project.StorageV1
+        - target_project.StorageV1
+      code: |-
+        def check(handles, max_score)
+          target_storage_handle = handles['target_project.StorageV1']
```

Comment:
do we want to pass handles in this format, or more like ```handles[:target_project][:StorageV1]```?
also we will probably need to list the possible services for each resource type

Reply:
i'd be fine with passing handles in either format - what do you think?
and yeah I have it as a "TODO" in the PR description to enumerate the services for each resource type.

Reply:
Is changing the contract of the arguments to this method in scope for a v2 bundle change? Alternatively, could we make a metadata change for all of the other fields and then talk about how we want the code snippet to function?

Reply:
yes, changing the contract of the arguments is in scope - that's actually the main reason for taking this on now: that multi-resource labs necessarily need a different signature

Reply:
TODO: Agreed. Why not both?